From 76632e2b3a9fd56290534a6be8543e265039fd22 Mon Sep 17 00:00:00 2001
From: Jami Kettunen <jami.kettunen@protonmail.com>
Date: Sun, 24 Nov 2024 17:30:41 +0200
Subject: [PATCH 15/15] [REVISIT] Make Mapplauncherd support optional

It's worth revisiting if this makes sense to keep as a dependency with
some repeatable benchmarks where a quantifiable difference may or may
not be observed when running this on a watch.

If this is to be removed cmake/FindMapplauncherd_qt*.cmake additionally
should also be gone.
---
 CMakeLists.txt          |  5 ++++-
 src/app/CMakeLists.txt  | 10 +++++++---
 src/app/asteroidapp.cpp | 14 ++++++++++++++
 3 files changed, 25 insertions(+), 4 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 3e06249..cae8606 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -9,6 +9,7 @@ find_package(ECM REQUIRED NO_MODULE)
 set(CMAKE_MODULE_PATH ${ECM_MODULE_PATH} ${CMAKE_SOURCE_DIR}/cmake)
 
 option(WITH_ASTEROIDAPP "Build the AsteroidApp class" ON)
+option(WITH_MAPPLAUNCHERD "Build the AsteroidApp class with Mapplauncherd booster support" ON)
 option(WITH_CMAKE_MODULES "Install AsteroidOS CMake modules" ON)
 
 include(FeatureSummary)
@@ -22,7 +23,9 @@ set(ASTEROID_MODULES_INSTALL_DIR ${CMAKE_INSTALL_DATADIR}/asteroidapp/cmake)
 find_package(Qt6 ${QT_MIN_VERSION} COMPONENTS DBus Qml Quick Svg ShaderTools REQUIRED)
 if (WITH_ASTEROIDAPP)
     find_package(Mlite6 MODULE REQUIRED)
-    find_package(Mapplauncherd_qt6 MODULE REQUIRED)
+    if (WITH_MAPPLAUNCHERD)
+        find_package(Mapplauncherd_qt6 MODULE REQUIRED)
+    endif()
 endif()
 
 ecm_find_qmlmodule(QtQuick.VirtualKeyboard 2.1)
diff --git a/src/app/CMakeLists.txt b/src/app/CMakeLists.txt
index 3d7d37d..be2cbfe 100644
--- a/src/app/CMakeLists.txt
+++ b/src/app/CMakeLists.txt
@@ -9,12 +9,13 @@ target_link_libraries(asteroidapp
 		Qt6::Quick
 	PRIVATE
 		Mlite6::Mlite6
-		Mapplauncherd_qt6::Mapplauncherd_qt6)
+		$<$<BOOL:${WITH_MAPPLAUNCHERD}>:Mapplauncherd_qt6::Mapplauncherd_qt6>)
 
 set_target_properties(asteroidapp PROPERTIES
 	EXPORT_NAME AsteroidApp
 	SOVERSION ${PROJECT_VERSION_MAJOR}
-	VERSION ${PROJECT_VERSION})
+	VERSION ${PROJECT_VERSION}
+	COMPILE_DEFINITIONS $<$<BOOL:${WITH_MAPPLAUNCHERD}>:WITH_MAPPLAUNCHERD=1>)
 
 target_include_directories(asteroidapp PUBLIC
 	$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}>
@@ -34,9 +35,12 @@ install(FILES asteroidapp.prf
 	DESTINATION ${CMAKE_INSTALL_LIBDIR}/mkspecs/features
 	COMPONENT Devel)
 
+if (WITH_MAPPLAUNCHERD)
+	set(ASTEROIDAPP_PKGCONF_DEPS "qdeclarative6-boostable")
+endif()
 ecm_generate_pkgconfig_file(
 	BASE_NAME asteroidapp
-	DEPS qdeclarative6-boostable
+	DEPS ${ASTEROIDAPP_PKGCONF_DEPS}
 	FILENAME_VAR asteroidapp
 	DESCRIPTION ${PROJECT_DESCRIPTION}
 	INSTALL)
diff --git a/src/app/asteroidapp.cpp b/src/app/asteroidapp.cpp
index e4fa8b7..0eeea38 100644
--- a/src/app/asteroidapp.cpp
+++ b/src/app/asteroidapp.cpp
@@ -33,6 +33,7 @@
 #include <QFileInfo>
 #include <MDesktopEntry>
 #include <QTranslator>
+#if WITH_MAPPLAUNCHERD
 #include <mdeclarativecache/MDeclarativeCache>
 
 static QString applicationPath()
@@ -47,11 +48,16 @@ static QString applicationPath()
         return QCoreApplication::applicationFilePath();
     }
 }
+#endif
 
 namespace AsteroidApp {
     QString appName()
     {
+#if WITH_MAPPLAUNCHERD
         QFileInfo exe = QFileInfo(applicationPath());
+#else
+        QFileInfo exe = QFileInfo(QCoreApplication::applicationFilePath());
+#endif
         return exe.baseName();
     }
 
@@ -60,7 +66,11 @@ namespace AsteroidApp {
         static QGuiApplication *app = NULL;
 
         if (app == NULL) {
+#if WITH_MAPPLAUNCHERD
             app = MDeclarativeCache::qApplication(argc, argv);
+#else
+            app = new QGuiApplication(argc, argv);
+#endif
 
             app->setOrganizationName(appName());
             app->setOrganizationDomain(appName());
@@ -80,7 +90,11 @@ namespace AsteroidApp {
 
     QQuickView *createView()
     {
+#if WITH_MAPPLAUNCHERD
         QQuickView *view = MDeclarativeCache::qQuickView();
+#else
+        QQuickView *view = new QQuickView;
+#endif
         MDesktopEntry entry("/usr/share/applications/" + appName() + ".desktop");
         if (entry.isValid()) {
             view->setTitle(entry.name());
-- 
2.47.0

